src/common.cpp

```diff
@@ -1849,61 +1856,122 @@ bool valid_func_name(const wcstring &str) {
 #else
     int name[] = {CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1};
 #endif
-    int result = sysctl(name, sizeof(name) / sizeof(int), buff, &buff_size, nullptr, 0);
-    if (result != 0) {
+    int ret = sysctl(name, sizeof(name) / sizeof(int), buff, &buff_size, nullptr, 0);
+    if (ret != 0) {
         wperror(L"sysctl KERN_PROC_PATHNAME");
     } else {
-        return std::string(buff);
+        char path[PATH_MAX];
+        if (realpath(buff, path)) {
+            result = path;
+            return result;
+        }
     }
-#else
-    // On other unixes, fall back to the Linux-ish /proc/ directory
+    result = "fish";
+    return result;
+#endif
     ssize_t len;
     len = readlink("/proc/self/exe", buff, sizeof buff - 1);  // Linux
     if (len == -1) {
-        len = readlink("/proc/curproc/file", buff, sizeof buff - 1);  // other BSDs
-        if (len == -1) {
-            len = readlink("/proc/self/path/a.out", buff, sizeof buff - 1);  // Solaris
-        }
+        len = readlink("/proc/self/path/a.out", buff, sizeof buff - 1);  // Solaris
     }
     if (len > 0) {
         buff[len] = '\0';
         // When /proc/self/exe points to a file that was deleted (or overwritten on update!)
         // then linux adds a " (deleted)" suffix.
         // If that's not a valid path, let's remove that awkward suffix.
-        std::string buffstr{buff};
+        buffstr = buff;
         if (access(buff, F_OK)) {
             auto dellen = const_strlen(" (deleted)");
             if (buffstr.size() > dellen &&
                 buffstr.compare(buffstr.size() - dellen, dellen, " (deleted)") == 0) {
                 buffstr = buffstr.substr(0, buffstr.size() - dellen);
             }
         }
-        return buffstr;
+        result = buffstr;
+        return result;
+    }
+    std::string strargv0 = ((argv0) ? argv0 : "");
+    struct stat st;
+    if (!strargv0.empty()) {
+        if (strargv0[0] == '/') {
+            buffstr = strargv0;
+        } else if (strargv0.find('/') == std::string::npos) {
+            const char *penv = getenv("PATH");
+            if (penv && *penv) {
+                std::vector<std::string> env = split_string(penv, ':');
+                for (std::size_t i = 0; i < env.size(); i++) {
+                    buffstr = env[i] + "/" + strargv0;
+                    if (!stat(buffstr.c_str(), &st) && (st.st_mode & S_IXUSR) && (st.st_mode & S_IFREG)) {
+                        break;
+                    }
+                }
+            }
+        } else {
+            const char *pwd = getenv("PWD");
+            if (pwd && *pwd) {
+                buffstr = std::string(pwd) + "/" + strargv0;
+            }
+        }
+    }
+    buffstr.resize(buff_size, '\0');
+    buffstr[buff_size] = '\0';
+    if (realpath(buffstr.c_str(), buff)) {
+        buffstr = buff;
     }
+#ifdef __OpenBSD__
+    // kd variable is necessary for kvm_getfiles() first argument; RTFM.
```

Comment:
It's obvious that the "kd" is passed in the next function, and keeping some "handle" is a fairly common idiom, no need to insult the reader.

Reply:
I wasn't sure if RTFM meant "Read the f'ing manual" or "Read the fine manual" -- I've heard it used both ways, I was hoping it was the latter.

Reply:
Both have been used and so people reading this might understand it either way, which means it's potentially insulting.

Anyway, the comment isn't very helpful - you can see that "kd" is used again a few lines later - so I'd just remove it.

But see the next comment, I'm not actually sold on the kvm_X stuff at all.